modules/custom/az_eds/az_eds_user/src/EventSubscriber/AzLdapCas.php

```diff
@@ -0,0 +1,182 @@
+    $original_filter = $this->ldapQuery->getFilter();
+    $filter = '(&' . $original_filter .
+      '(uid=' .
+      ldap_escape($authname, "", LDAP_ESCAPE_FILTER) .
+      '))';
+    $this->ldapQuery->execute($filter);
+    $results = $this->ldapQuery->getRawResults();
+    foreach ($results as $result) {
+      // Make sure the result contains the user.
+      $row = $result->getAttributes();
+      $uid = $row['uid'] ?? [];
+      $uid = reset($uid);
+      if ($uid === $authname) {
+        return TRUE;
+      }
+    }
+    return FALSE;
+  }
+
+  /**
+   * Respond to events on CAS validation for JIT provisioning.
+   *
+   * @param \Drupal\cas\Event\CasPostValidateEvent $event
+   *   The event object.
+   */
+  public function onValidate(CasPostValidateEvent $event) {
+    // See who was validated.
+    $bag = $event->getCasPropertyBag();
+    $username = $bag->getUsername();
+    // Check if a CAS user exists.
+    $cas_uid = $this->externalAuth->getUid($username, 'cas');
+    if ($cas_uid === FALSE) {
+      if (!$this->userAllowedByQuery($username)) {
+        // Not allowed by query.
+        return;
+      }
+      // User does not exist, but is allowed, attempt LDAP provisioning.
+      $result = $this->drupalUserProcessor->createDrupalUserFromLdapEntry(
+        [
+          'name' => $username,
+          'status' => TRUE,
+        ]
+      );
+      // Successfully provisioned LDAP user for which there is no cas account.
+      if ($result) {
+        // Get the mapped uid of the new user.
+        if ($ldap_uid = $this->externalAuth->getUid($username, 'ldap_user')) {
+          $user = $this->entityTypeManager->getStorage('user')->load($ldap_uid);
+          // We have the user that ldap_user provisioned, set the cas account.
+          if (!empty($user)) {
+            $this->casUserManager->setCasUsernameForAccount($user, $user->getAccountName());
+          }
+        }
+      }
+    }
+    // CAS user existed but disallowed by query (e.g. user has lost membership.)
+    elseif (!$this->userAllowedByQuery($username)) {
+      if ($ldap_uid = $this->externalAuth->getUid($username, 'ldap_user')) {
+        $user = $this->entityTypeManager->getStorage('user')->load($ldap_uid);
+        // User originally provisioned by LDAP, therefore cancel the account.
+        // The implication is to NOT cancel an account created by hand.
+        // @todo is this workflow correct?
+        if (!empty($user) && !$user->isBlocked()) {
+          // Deactivate the account.
+          $user->block();
+          $user->save();
+          // After this event, CAS login will fail because account blocked.
+        }
+      }
+    }
+    // CAS user existed, and is still allowed by query.
+    else {
+      $user = $this->entityTypeManager->getStorage('user')->load($cas_uid);
+      // Check if user needs to be unblocked.
+      if (!empty($user) && $user->isBlocked()) {
+        // Deactivate the account.
+        $user->activate();
```

Comment:
This reactivation should likely only occur if they were an LDAP-provisioned user.